### ISSUE TITLE
UQ Methods to add

### ISSUE THREAD

**Issue by nilsleh** [lightning-uq-box/lightning-uq-box#43]
> [!NOTE]  
> If you have a request to support a specific method, or would like to see priority of one of the listed methods, please open a separate issue, so it won't get buried in this thread. Based on the issue, we can then add it to the overall list. Thanks and looking forward to your ideas, requests and contributions.

This issue can be a collection and discussion of methods we could add to the library at some point, in no particular order :) 

An overview of some methods from the survey paper by [Gawlikowski et al. 2023](https://link.springer.com/article/10.1007/s10462-023-10562-9) is [here](https://github.com/JakobCode/UncertaintyInNeuralNetworks_Resources).

## Methods remaining to be grouped
- [Stein Variational Gradient Descent](https://arxiv.org/abs/1608.04471)
- [Density Uncertainty Layers](https://arxiv.org/pdf/2306.12497.pdf) with [code](https://github.com/yookoon/density_uncertainty_layers)
- [Sampling from Gaussian Process Posteriors using Stochastic Gradient Descent](https://arxiv.org/abs/2306.11589)
- [CARD: Classification and Regression Diffusion Models](https://arxiv.org/abs/2206.07275) (Implemented in #47) but missing support for classification
- [Epistemic Neural Networks](https://arxiv.org/abs/2107.08924)
- [Prediction-powered inference](https://arxiv.org/pdf/2301.09633.pdf)
- [Collapsed Inference for Bayesian Deep Learning](https://github.com/ucla-starai/ciber)
- [Posterior Refinement Improves Sample Efficiency in BNNs](https://github.com/runame/laplace-refinement)
- [Quantification of Uncertainty with Adverserial Models](https://arxiv.org/pdf/2307.03217.pdf)
- [Single Model Uncertainty](https://github.com/facebookresearch/SingleModelUncertainty)
- [Repulsive Deep Ensembles](https://github.com/ratschlab/repulsive_ensembles)
- [Neural Diffusion Processes](https://github.com/vdutor/neural-diffusion-processes)
- [DropConnect](https://arxiv.org/pdf/1906.04569.pdf)
- [Probabilistic Circuits](https://github.com/ml-research/tractable-dropout-inference)
- [Single-shot Bayesian Approximation for NN](https://arxiv.org/abs/2308.12785) [Code](https://github.com/kaibrach/Moment-Propagation)
- [Implicit Variational Inference for High-Dimensional Posteriors](https://arxiv.org/abs/2310.06643)
- [Convolutional Deep Kernel Machine](https://arxiv.org/abs/2309.09814)
- [Spline Prediction Intervals via Conformal Estimation (SPICE)](https://github.com/ndiamant/spice)
- [Evidential Deep Learning: Predictive Uncertainty for Earth System Modelling](https://arxiv.org/abs/2309.13207)
- [Uncertainty-aware Surrogate Models for Airfoil Flow Simulations with Denoising Diffusion Probabilistic Models](https://github.com/tum-pbs/Diffusion-based-Flow-Prediction)
- [SNGP](https://arxiv.org/abs/2006.10108) #98 
- [Direct Epistemic Uncertainty Prediction (DEUP)](https://github.com/MJ10/DEUP)
- [MixupMP](https://arxiv.org/abs/2403.12729)
- [EasyUQ](https://github.com/evwalz/easyuq)
- [Lightweight Probabilistic Networks](https://github.com/mattiasegu/uncertainty_estimation_deep_learning)
- [UQ via Distribution Propagation](https://www.arxiv.org/pdf/2402.08324)
- [Rate-in: Adaptive Dropout](https://arxiv.org/abs/2412.07169) with [Code](https://github.com/code-supplement-25/rate-in/tree/main)

## Methods by Task

Regression:
- [Deep Evidential Regression for Bayesian Quantile Regression](https://arxiv.org/abs/2308.10650)
- [Autoregressive Conditional Neural Process](https://arxiv.org/pdf/2303.14468.pdf)
- [Convolutional Conditional Neural Process](https://github.com/wesselb/neuralprocesses)
- [Wasserstein Generative Regression](https://arxiv.org/abs/2306.15163)
- [Escaping the Sampling Trap: Fast and Accurate Epistemic Uncertainty Estimation with Pairwise-Distance Estimators](https://openreview.net/forum?id=JdWpIe70FL)
- [Variational Bayesian Last Layer](https://openreview.net/forum?id=Sx7BIiPzys&referrer=%5Bthe%20profile%20of%20Jasper%20Snoek%5D(%2Fprofile%3Fid%3D~Jasper_Snoek1)) #137 
- [End to End Conformal](https://github.com/chrisyeh96/e2e-conformal)

Classification:
- [DropMax](https://github.com/zsef123/DropMax-pytorch/blob/master/DropMax.py)
- [Deterministic Uncertainty Quantification (DUQ)](https://github.com/y0ast/deterministic-uncertainty-quantification#deterministic-uncertainty-quantification-duq)
- [Deep Deterministic Uncertainty](https://github.com/omegafragger/DDU)
- [Deep Evidential Classification](https://github.com/dougbrion/pytorch-classification-uncertainty)
- [Unified Uncertainty Calibration](https://github.com/facebookresearch/UnifiedUncertaintyCalibration)
- [Dirichlet Based Uncertainty Models](https://github.com/TUM-DAML/dbu-robustness)
- [Single Layer Predictive Normalized OOD](https://github.com/kobybibas/pnml_ood_detection)
- [Deep Kernel Machines](https://github.com/edwardmilsom/skr_cdkm)

PixelWise Regression:
- [BayesCap](https://github.com/ExplainableML/BayesCap/tree/main)
- [Marigold](https://github.com/prs-eth/marigold) possibly interesting EO application, intended for Monocular Depth Estimation
- [Image2Image Regression](https://arxiv.org/abs/2202.05265) with [Code](https://github.com/aangelopoulos/im2im-uq) #118 
- [Deep Energy Based Models](https://uvadlc-notebooks.readthedocs.io/en/latest/tutorial_notebooks/tutorial8/Deep_Energy_Models.html)
- [Normalizing Flows](https://uvadlc-notebooks.readthedocs.io/en/latest/tutorial_notebooks/tutorial11/NF_image_modeling.html)

Segmentation:
- [Probabilistic Unet](https://github.com/stefanknegt/Probabilistic-Unet-Pytorch)
- [Segmentation Diffusion](https://github.com/tomeramit/SegDiff)
- [Deep Deterministic Uncertainty](https://arxiv.org/abs/2111.00079)

Object Detection:
- [Adaptive Bounding Box Conformal](https://github.com/alextimans/conformal-od)

Generative Models:
- [LatentDiffusion Model](https://github.com/CompVis/latent-diffusion) and with [UQ application](https://github.com/MeteoSwiss/ldcast)

Post-hoc calibration procedures:
- [RASP Conformal](https://github.com/aangelopoulos/conformal_classification) (#57)
- [Temperature Scaling](https://github.com/gpleiss/temperature_scaling) (#56)
- [Conformal Risk Control](https://github.com/aangelopoulos/conformal-risk)
- [Conditional Conformal Prediction](https://github.com/jjcherian/conditional-conformal)
- [SASP Conformal](https://arxiv.org/pdf/2310.06430.pdf)
- [MC-Dropout Conformal](https://github.com/team-daniel/MC-CP/blob/master/AAAISupplementaryCode.ipynb)
- [Monte Carlo Conformal](https://github.com/google-deepmind/uncertain_ground_truth)

**Comment by wrkhard** [comment]
I would be very interested in seeing SNGP implemented for ViTs. Seems to be very few UQ methods out there for ViT.

Paper: https://papers.nips.cc/paper/2020/file/543e83748234f7cbab21aa0ade66565f-Paper.pdf
TF implementation for ResNet: https://www.tensorflow.org/tutorials/understanding/sngp#full-sngp-model

**Comment by qqaazz800624** [comment]
Thanks for your hard work! 
I just came across a 2024 ICLR paper a few days ago which I found it quite interesting and encouraging. The paper proposed BayesDiff, a pixel-wise uncertainty estimator for generations from diffusion models based on Bayesian inference. 
Since Diffusion models has gained so much attention and proved to have impressive image generation capability, I am glad to know that someone has found a way to combine uncertainty quantification and Diffusion models. 

<img width="1112" alt="截圖 2024-04-22 下午5 14 03" src="https://github.com/lightning-uq-box/lightning-uq-box/assets/79075772/7c47bc4d-7b37-468f-949a-b3e7cedf7a21">

KOU, Siqi, et al. BayesDiff: Estimating Pixel-wise Uncertainty in Diffusion via Bayesian Inference. arXiv preprint arXiv:2310.11142, 2023.

The Github:
https://github.com/karrykkk/BayesDiff

**Comment by wrkhard** [comment]
Hi @nilsleh 

I was wondering if you will be adding EMA support to CARD regression. Saw a "TODO" buried in the code. Also very much looking forward to your Lightning implementation of Neural Diffusion Processes! Thanks again for all of your amazing work!

**Comment by nilsleh** [comment]
> I was wondering if you will be adding EMA support to CARD regression. Saw a "TODO" buried in the code. Also very much looking forward to your Lightning implementation of Neural Diffusion Processes! Thanks again for all of your amazing work!

Hi @wrkhard , could you open a separate issue for this, I am afraid I will forget your request otherwise (will add a note on top for future). On a different note, I am currently spending a lot of time in implementing a Diffusion model as a base for various UQ methods around it. There I am adding EMA, so I expect it to not be difficult to support it for CARD as well. Just cannot make promises about a timeline, sorry.
